client/src/main/scala/org/apache/celeborn/client/RequestLocationCallContext.scala

```diff
@@ -17,23 +17,56 @@
 
 package org.apache.celeborn.client
 
+import java.util
+import java.util.concurrent.ConcurrentHashMap
+
+import org.apache.celeborn.common.internal.Logging
 import org.apache.celeborn.common.protocol.PartitionLocation
 import org.apache.celeborn.common.protocol.message.ControlMessages.{ChangeLocationResponse, RegisterShuffleResponse}
 import org.apache.celeborn.common.protocol.message.StatusCode
 import org.apache.celeborn.common.rpc.RpcCallContext
 
 trait RequestLocationCallContext {
-  def reply(status: StatusCode, partitionLocationOpt: Option[PartitionLocation]): Unit
+  def reply(
+      partitionId: Int,
+      status: StatusCode,
+      partitionLocationOpt: Option[PartitionLocation]): Unit
 }
 
-case class ChangeLocationCallContext(context: RpcCallContext) extends RequestLocationCallContext {
-  override def reply(status: StatusCode, partitionLocationOpt: Option[PartitionLocation]): Unit = {
-    context.reply(ChangeLocationResponse(status, partitionLocationOpt))
+case class ChangeLocationsCallContext(
+    context: RpcCallContext,
+    partitionCount: Int)
+  extends RequestLocationCallContext with Logging {
+  val endedMapIds = new util.HashSet[Integer]()
+  val newLocs = new ConcurrentHashMap[Integer, (StatusCode, PartitionLocation)](partitionCount)
+
+  def markMapperEnd(mapId: Int): Unit = this.synchronized {
+    endedMapIds.add(mapId)
+  }
+
+  override def reply(
+      partitionId: Int,
+      status: StatusCode,
+      partitionLocationOpt: Option[PartitionLocation]): Unit = this.synchronized {
+    newLocs.putIfAbsent(
```

Comment:
It can be replaced by computeIfAbsent. 

[computeIfAbsent](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#computeIfAbsent-K-java.util.function.Function-) takes a mapping function, that is called to obtain the value if the key is missing.

[putIfAbsent](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#putIfAbsent-K-V-) takes the value directly.

If the value is expensive to obtain, then putIfAbsent wastes that if the key already exists.

A common "expensive" value is e.g. new ArrayList<>() for when you're creating a Map<K, List<V>>, where creating a new list when the key already exists (which then discards the new list) generates unnecessary garbage.

Reply:
> It can be replaced by computeIfAbsent.
> 
> [computeIfAbsent](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#computeIfAbsent-K-java.util.function.Function-) takes a mapping function, that is called to obtain the value if the key is missing.
> 
> [putIfAbsent](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#putIfAbsent-K-V-) takes the value directly.
> 
> If the value is expensive to obtain, then putIfAbsent wastes that if the key already exists.
> 
> A common "expensive" value is e.g. new ArrayList<>() for when you're creating a Map<K, List>, where creating a new list when the key already exists (which then discards the new list) generates unnecessary garbage.

Thanks for the comment. In fact it should not happen that duplicate partitionIds are replied here. In addition, I the synchronized is not necessary here.